docs/util/build-error-docs.py

```diff
@@ -0,0 +1,151 @@
+    return anchor
+
+
+def generate_doc_rows(class_name, class_details):
+    class_row = [
+        """
+        <tr id="{anchor}">
+            <td>{sql_state}</td>
+            <td>
+                <span class="error-name">
+                    <code>
+                    <a href="#{anchor}">#</a>
+                    </code>
+                    {class_name}
+                </span>
+            </td>
+            <td>{message}</td>
+        </tr>
+        """
+        .format(
+            anchor=anchor_name(class_name),
+            sql_state=class_details["sqlState"],
+            # This inserts soft break opportunities so that if a long name needs to be wrapped
+            # it will wrap in a visually pleasing manner.
+            # See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/wbr
+            class_name=class_name.replace("_", "<wbr />_"),
+            message=class_details["message"],
+        )
+    ]
+    sub_class_rows = []
+    if "subClass" in class_details:
+        for sub_class_name in sorted(class_details["subClass"]):
+            sub_class_rows.append(
+                """
+                <tr id="{anchor}">
+                    <td></td>
+                    <td class="error-sub-class">
+                        <span class="error-name">
+                            <code>
+                            <a href="#{anchor}">#</a>
+                            </code>
+                            {sub_class_name}
+                        </span>
+                    </td>
+                    <td class="error-sub-class">{message}</td>
+                </tr>
+                """
+                .format(
+                    anchor=anchor_name(class_name, sub_class_name),
+                    # See comment above for explanation of `<wbr />`.
+                    sub_class_name=sub_class_name.replace("_", "<wbr />_"),
+                    message=class_details["subClass"][sub_class_name]["message"],
+                )
+            )
+    doc_rows = class_row + sub_class_rows
+    return [
+        dedent(row).strip()
+        for row in doc_rows
+    ]
+
+
+def generate_doc_table(error_classes):
+    doc_rows = chain.from_iterable([
+        generate_doc_rows(class_name, class_details)
+        for class_name, class_details
+        in sorted(
+            error_classes.items(),
+            key=lambda x: (x[1]["sqlState"], x[0]),
+        )
+    ])
+    table_html = (
+        """
+        <table id="error-conditions">
+        <tr>
+            <th>Error State / SQLSTATE</th>
+            <th>Error Condition & Sub-Condition</th>
+            <th>Message</th>
+        </tr>
+        {rows}
+        </table>
+        """
+    )
+    # We dedent here rather than above so that the interpolated rows (which are not
+    # indented) don't prevent the dedent from working.
+    table_html = dedent(table_html).strip().format(rows="\n".join(list(doc_rows)))
+    return table_html
+
+
+if __name__ == "__main__":
```

Comment:
when do we run this script? Now the generated doc pages are not committed in the codebase.

Reply:
Correct, this is the main goal of this PR: to remove auto-generated content from git, plus all the associated friction it causes.

One of my [TODOs for this PR][1] is to invoke this script from the Jekyll build. In other words, no-one will need to remember to run anything separately. They just run `bundle exec jekyll build` as they normally do.

[1]: https://github.com/apache/spark/pull/44971#issue-2111332887